### ISSUE TITLE
How to Save/Load state

### ISSUE THREAD

**Issue by JLLMorales** [inkle/ink#374]
Hi,

I cannot achieve to save/load state of my ink stories. I have used inky to write my story, and integrated it in Unity 3d, with the blot example. Everything works fine, but I don't know how to save/load the state of the game. 
In the manual says:

_Saving and loading
To save the state of your story within your game, call:
string savedJson = _inkStory.state.ToJson();
...and then to load it again:
_inkStory.state.LoadJson(savedJson);_

But I don't know what to do with it. Do I have to create a script? 

Sorry, I don't know anything about programming, but it's the last thing I need to finish my project and don't make the player to start all over again everytime he runs the game.

Thank you very much.

JL Lopez Morales

**Comment by joethephish** [comment]
Yep, I'm afraid you need to do a bit of scripting. I'd suggest saving the state using Unity's [PlayerPrefs](https://docs.unity3d.com/ScriptReference/PlayerPrefs.html).

So, for example, when you first start the game (e.g. when you create your ink story object), try to load a saved game like this:

```
if( PlayerPrefs.HasKey("inkSaveState") ) {
    var savedState = PlayerPrefs.GetString("inkSaveState");
    inkStory.state.LoadJson(savedState);
}
```

Then, every time the player makes a choice, save the state:

```
var savedState = _inkStory.state.ToJson();
PlayerPrefs.SetString("inkSaveState", savedState);
```

You'll also need a way to clear the saved game though, in case the player wants to start again from scratch, or perhaps if you get to an ending. To do that, you can clear the `PlayerPrefs`, as well as reset the ink story:

```
PlayerPrefs.DeleteKey("inkSaveState");
_inkStory.ResetState()
```

**Comment by JLLMorales** [comment]
Thank you very much for your answer. I will try to do as you say, even I didn't used before the PlayerPrefs. it's a pity that inky doen't have a save/load function, so everything would be easier.

**Comment by joethephish** [comment]
External functions aren't a good idea in this situation since they're called in the middle of the ink running, which would be dangerous and probably wouldn't work. The best way is to simply have some special pure text in your story. For example:

```
== new_chapter ==
CHECKPOINT
You arrive in a cave.
* Start from last checkpoint -> back_to_checkpoint
* Start over -> start_over


== start_over
ENDING
-> END

== back_to_checkpoint
BACK TO CHECKPOINT
-> END
```

Where you make your calls to `story.Continue()` etc, add a thing that checks for the specific strings `CHECKPOINT`, `ENDING`, and `BACK TO CHECKPOINT`. And if it sees one, call your C# specific code.

This is the best way since the ink engine is in a stable state (not in the middle of evaluating) at the point where you read those strings in.

**Comment by Miha-TzarG** [comment]
This Load  is work for me

string savedJson;
void Start(){ // or your start for story
		if (PlayerPrefs.HasKey("inkSaveState"))
		{

			story = new Story(inkJSONAsset.text);
			savedJson = PlayerPrefs.GetString("inkSaveState");
			story.state.LoadJson(savedJson);
			Debug.Log(savedJson);
			RefreshView(); // or   refresh();
		}
		else
		{

			story = new Story(inkJSONAsset.text);
                       //RefreshView();
		
		}

}
save :

savedJson = story.state.ToJson();
				PlayerPrefs.SetString("inkSaveState", savedJson);
